lib/active_model/serializer/array_serializer.rb

```diff
@@ -21,11 +22,11 @@ def initialize(objects, options = {})
       end
 
       def json_key
-        @objects.first.json_key if @objects.first
-      end
-
-      def root=(root)
-        @objects.first.root = root if @objects.first
+        if root == true && @objects.first
+          @objects.first.class.root_name
```

Comment:
The problem here is the jsonapi adapter.  It sets `serializer.root = true` in its initializer.  I'm not sure how to deal with heterogenous arrays.